scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+                              self.samples_per_frame)
+        else:
+            d_shape_header = (self.nbin, self.nchan, self.npol)
+        if d_shape_raw != (self.nrow, ) + d_shape_header[::-1]:
+            raise ValueError("Data shape does not match with the header"
+                             " information")
+        if self.mode == "SEARCH":
+            d_shape = namedtuple('d_shape', ['nrow', 'nsample', 'npol', 'nchan',
+                                             'nbin'])
+            result  = d_shape(self.nrow,  self.samples_per_frame, self.npol,
+                              self.nchan, self.nbin)
+        else:
+            d_shape = namedtuple('d_shape', ['nrow', 'npol', 'nchan', 'nbin'])
+            result  = d_shape(self.nrow,  self.npol, self.nchan, self.nbin)
+        return result
+
+    @property
+    def polarization(self):
+        return self.header['POL_TYPE']
+
+    @polarization.setter
+    def polarization(self, val):
+        self.header['POL_TYPE'] = val
+
+    @property
+    def frequency(self):
+        if 'DAT_FREQ' in self.columns.names:
+            freqs = u.Quantity(self.data['DAT_FREQ'],
+                               u.MHz, copy=False)
+            assert np.isclose(freqs, freqs[0],
+                              atol=np.finfo(float).eps * u.MHz).all(), \
+                "Frequencies are different within one subint rows."
+            freqs = freqs[0]
+        else:
+            # Get the frequency from the header HDU
+            try:
+                freqs = self.header_hdu.frequency
+            except:
+                freqs = None
+        return freqs
+
+    @property
+    def dtype(self):
+        return self.data['DATA'].dtype
+
+    @frequency.setter
+    def frequency(self, val):
+        if 'DAT_FREQ' in self.columns.names:
+            self.data['DAT_FREQ'] = val
+            ## Check frequency length.
+            assert self.shape.nchan == len(val), \
+                "Input frequency does not match the number of channels in data."
+        else:
+            self._req_columns['DAT_FREQ']['value'] = val
+
+    def read_data_row(self, row_index):
+        if row_index >= self.shape[0]:
+            raise EOFError("cannot read from beyond end of input SUBINT HDU.")
+
+        row = self.data[row_index]
+        # Reversed the header shape to match the data
+        new_shape = self.raw_shape._replace(samples_per_frame=1,
+                                            nbin=1)[-1:1:-1]
+        data_scale = row['DAT_SCL'].reshape(new_shape)
+        data_off_set = row['DAT_OFFS'].reshape(new_shape)
+        zero_off = np.asarray(0, dtype=self.dtype)
+        if 'ZERO_OFF' in self.header.keys():
+            try:
+                zero_off = np.asarray(self.header['ZERO_OFF'], dtype=self.dtype)
+            except:
+                pass
+        result = ((row['DATA'] - zero_off)* data_scale +
+                   data_off_set)
+        data_wts = np.ones(self.nchan)
```

Comment:
Instead of doing nothing, I'd just put the multiplication with `data_wts` inside the `if` clause.